### PR TITLE
Initialize using Dependency Set Configuration

### LINKED ISSUE #12487: Interop: use dependency-set to define chain sub-services
Make the dependency-set iterable, so we can instantiate the necessary sub-services per chain.

Sub-services per chain:
- a `fromda` DB, added to the `ChainsDB` bundle
- a `logs` DB, added to the `ChainsDB` bundle
- a cross-unsafe updater (interface for now, being implemented in #12075 )
- a cross-safe updater (interface for now, like the above)

The `AddRPC` should add a chain-processor only, and abort if:
- one already exist for the given chain (we don't currently support load-balancing between multiple RPC sources yet)
- one is not supposed to exist (i.e. chain of the RPC is not part of the dependency set)
Update: attaching the RPC, to a previously initialized processor, seems like a better flow.

### PR DESCRIPTION
**Pull request by axelKingsley**
# What
Uses the new Dependency Set Configuration to spin up database constructs for exactly the chains specified.

# How
I split out the "AddL2FromRPC" functionality into two parts:

### openChainDBs

Creates the LogDB and FromDA DBs for the given chain and attaches it to the `SupervisorBackend.chainsDB`.

### ~~addProcessorFromRPC~~ attachRPC

Connects via RPC to determine ChainID, and then attaches a new processor to `SupervisorBackend.ChainProcessors`.
If the discovered ChainID is not part of the dependency set (determined by known processors), this attachment fails.

Chain-Processors are now instantiated upfront, to allow attaching of RPCs during runtime, and to make the test-setup a little more sane (we can attach a mock source with `AttachProcessorSource` now)

### Additionally
The `SupervisorBackend` now tracks a map for `ChainMetrics`, as both the Processor and Database want to use it, and the processor may not be attached until later.

Now, when users use the `AddRPC` function on the Supervisor, it *only* follows the `addProcessorFromRPC` path.

And a minor bugfix in the `logs` db, to read the full block seal properly.

# TODO:
* ~~Unit Tests~~ (done: there's a backend test asserting the presence of DBs and unsafe/cross-unsafe block functionality)
* ~~Creating the `fromda` objects inside the `ChainsDB` when the call is made~~ (done)

# Metadata

Fix https://github.com/ethereum-optimism/optimism/issues/12487

### CODE REVIEW

File: op-supervisor/supervisor/backend/backend.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/backend.go:42]: just changing the name of this to match the other "chainXYZ" structures this holds

File: op-supervisor/supervisor/backend/backend.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/backend.go:53]: chainMetrics are now held by the Backend so they may be referenced later by resources which get spun up later.

File: op-supervisor/supervisor/backend/backend.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/backend.go:127]: Every chain in the Supervisor is maintained by the backend in ~2 ways:
* Database Storage
* RPC Connection

The databases can be generated per chain as soon as we know about the chain, as the chainID is the only important part. However, the RPC connections are not always ready. We found this was an issue in testing, a circular bootstrap requirement where the Nodes and Supervisor must be aware of one-another.

So, now we make databases for *all* chains, and we connect to any RPCs specified in the config. As we connect to them, we discover which chain they are for and connect them. And you can use the AddRPC API to add more connections later.

These RPC connections are how the Supervisor gets logs. The Nodes push updates ("I have a new Unsafe block"), and the Supervisor will respond by pulling the receipts for that block using the RPC for that chain.

@protolambda: note, there's an inconsistency risk here, where the node informing you of an update might *not* be the node you connect to via RPC. you could discover that the RPC node hasn't gotten the block yet. For now we can solve this by just saying "don't do that, use one node only".

File: op-supervisor/supervisor/backend/backend.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/backend.go:133]: When I mentioned that there's a database per chain, there's actually multiple:
* LogsDB, where the raw log data is stored
* Local Derived From, where L1 derivation information is stored
* Cross Derived From, where more L1 derivation information is held

File: op-supervisor/supervisor/backend/backend.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/backend.go]:
```suggestion
```

File: op-supervisor/supervisor/backend/depset/depset.go
Comment (axelKingsley) [inline op-supervisor/supervisor/backend/depset/depset.go:34]: This *could* be excluded from the interface, callers to `Chains` can already decide this, but this way we save on having to copy out the whole list every time.

**Review by axelKingsley — COMMENTED** [review]
I'm the author of this PR so I can't approve it, but since @protolambda added some things, I'm going to review/annotate things to make review as easy as possible for @tyler-smith.

@protolambda generally in favor of the changes you added.
* I see you removed `checkProcessorCoverage`, first taking out the error return, then removing the function altogether. I think that function is pretty helpful because it can signal to callers that the backend isn't properly initialized yet, and the log messages would be very nice to see if ever a supervisor is failing due to incomplete coverage.
* Looks like you put together a different synchronization mechanism for the Processors. That's fine, but I'm not sure it really belonged in this PR.

I would approve this if I wasn't the author :)

**Review by protolambda — APPROVED** [review]

**Pull request merged**
